core/src/main/java/ai/timefold/solver/core/impl/bavet/common/RecordAndReplayPropagator.java

```diff
@@ -87,33 +84,51 @@ public void retract(Object object) {
     @Override
     public void propagateRetracts() {
         if (!retractQueue.isEmpty() || !insertQueue.isEmpty()) {
+            var precomputeBuildHelper = precomputeBuildHelperSupplier.get();
+            var internalNodeNetwork = precomputeBuildHelper.getNodeNetwork();
+            var recordingTupleLifecycle = precomputeBuildHelper.getRecordingTupleLifecycle();
+
             updateQueue.removeAll(retractQueue);
             updateQueue.removeAll(insertQueue);
+
+            var allObjectSet = CollectionUtils
+                    .newIdentityHashSet(insertQueue.size() + objectToOutputTuplesMap.size());
+
             // Do not remove queued retracts from inserts; if a fact property
             // change, there will be both a retract and insert for that fact
+
+            allObjectSet.addAll(objectToOutputTuplesMap.keySet());
+            allObjectSet.removeAll(retractQueue);
+            allObjectSet.addAll(insertQueue);
+
             invalidateCache();
+            retractQueue.forEach(objectToOutputTuplesMap::remove);
 
-            retractQueue.forEach(this::retractFromInternalNodeNetwork);
-            insertQueue.forEach(this::insertIntoInternalNodeNetwork);
             retractQueue.clear();
             insertQueue.clear();
 
+            for (var object : allObjectSet) {
+                internalNodeNetwork.getRootNodesAcceptingType(object.getClass())
+                        .forEach(node -> ((BavetRootNode<Object>) node).insert(object));
+            }
```

Comment:
This is a major source of inefficiency.
The determination of which nodes accept the type happens over and over, every time we process one of the hundreds of thousands of objects.

If we already have to aggregate all the objects, then we might as well aggregate them into a map of sets, where the key is the class. In that way, the overhead of node processing will be much less. It may prevent us from properly sizing the sets, but I consider that only a small downside.